File: DataStructures/Disjoint Set Union/atcoder/A66/A66.php
<?php

fscanf(STDIN, "%d %d", $N, $Q);

// Union-Find クラス定義
class UnionFind
{
    private array $parent;
    private array $rank;

    public function __construct(int $n)
    {
        $this->parent = range(0, $n);  // 1-indexed
        $this->rank = array_fill(0, $n + 1, 0);
    }

    public function find(int $x): int
    {
        if ($this->parent[$x] !== $x) {
            $this->parent[$x] = $this->find($this->parent[$x]);  // 経路圧縮
        }
        return $this->parent[$x];
    }

    public function union(int $x, int $y): void
    {
        $rx = $this->find($x);
        $ry = $this->find($y);
        if ($rx === $ry) return;

        if ($this->rank[$rx] < $this->rank[$ry]) {
            $this->parent[$rx] = $ry;
        } else {
            $this->parent[$ry] = $rx;
            if ($this->rank[$rx] === $this->rank[$ry]) {
                $this->rank[$rx]++;
            }
        }
    }

    public function same(int $x, int $y): bool
    {
        return $this->find($x) === $this->find($y);
    }
}

$uf = new UnionFind($N);
$output = [];

for ($i = 0; $i < $Q; $i++) {
    fscanf(STDIN, "%d %d %d", $t, $u, $v);
    if ($t === 1) {
        $uf->union($u, $v);
    } else {
        $output[] = $uf->same($u, $v) ? "Yes" : "No";
    }
}

echo implode("\n", $output) . "\n";

File: DataStructures/Disjoint Set Union/atcoder/A66/A66.py
import sys
sys.setrecursionlimit(1 << 25)
input = sys.stdin.readline

class UnionFind:
    def __init__(self, n):
        self.parent = list(range(n + 1))  # 1-indexed
        self.rank = [0] * (n + 1)

    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])  # 経路圧縮
        return self.parent[x]

    def union(self, x, y):
        rx = self.find(x)
        ry = self.find(y)
        if rx == ry:
            return
        if self.rank[rx] < self.rank[ry]:
            self.parent[rx] = ry
        else:
            self.parent[ry] = rx
            if self.rank[rx] == self.rank[ry]:
                self.rank[rx] += 1

    def same(self, x, y):
        return self.find(x) == self.find(y)

def main():
    N, Q = map(int, input().split())
    uf = UnionFind(N)
    res = []
    for _ in range(Q):
        t, u, v = map(int, input().split())
        if t == 1:
            uf.union(u, v)
        else:
            res.append("Yes" if uf.same(u, v) else "No")
    print('\n'.join(res))

if __name__ == "__main__":
    main()
